Support new hubble metrics context: "labelsContext"

chancez opened the pull request:
The new "labelsContext" metrics context differs from the existing
contexts in that instead of putting different values into the
source/destination labels, it accepts a list of predefined labels that
should be exposed, and maps the flow data into these different labels.

This PR, we add source/destination namespace, pod, workload, and app
labels as well as a reporter label which allow distinguishing where the
metric was counted (client vs server). By adding this `reporter` label,
we can avoid potentially double counting metrics when using both egress
and ingress policies with the same port, since the metric will be
produced on both sides of the connection. For more details on the new
labels, see the updated documentation.

For comparison, Istio documents the metrics and labels it exposes:
https://istio.io/latest/docs/reference/config/metrics/.

This makes the hubble metrics much easier to use as it now separates out
the different values into separate labels.

Here is an example query that illistrates using these new metrics,
enabling getting the total number of requests by the destination
workload, filtered to metrics reported the server side (as opposed to
the client).

```
sum(rate(hubble_http_requests_total{reporter="server"}[1m])) by (destination_workload, method)
```

Release note:

```release-note
Support new hubble metrics context: "labelsContext"
```

chancez committed:
hubble/metrics: Support new metrics context: "labelsContext"

The new "labelsContext" metrics context differs from the existing
contexts in that instead of putting different values into the
source/destination labels, it accepts a list of predefined labels that
should be exposed, and maps the flow data into these different labels.

In this commit, we add source/destination namespace, pod, workload, and app
labels as well as a traffic_direction label which allow distinguishing where the
metric was counted (ingress vs egress). By adding this `traffic_direction` label,
we can avoid potentially double counting metrics when using both egress
and ingress policies with the same port, since the metric will be
produced on both sides of the connection. For more details on the new
labels, see the updated documentation.

For comparison, Istio documents the metrics and labels it exposes:
https://istio.io/latest/docs/reference/config/metrics/.

This makes the hubble metrics much easier to use as it now separates out
the different values into separate labels.

Here is an example query that illustrates using these new metrics,
enabling getting the total number of requests by the destination
workload, filtered to metrics reported the server side (as opposed to
the client).

```
sum(rate(hubble_http_requests_total{traffic_direction="ingress"}[1m])) by (destination_workload, method)
```

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -39,18 +39,33 @@ const (
 	// purpose. It uses "reserved:kube-apiserver" label if it's present in the identity label list.
 	// Otherwise, it uses the first label in the identity label list with "reserved:" prefix.
 	ContextReservedIdentity
+	// ContextReservedIdentity uses reserved labels in the identity label list for identification
```
Good eye. This code actually isn't needed anymore anyways.

chancez commented on Documentation/observability/metrics.rst:
```diff
@@ -530,6 +530,29 @@ and use the IP address of the target instead.
+========================= ===============================================================================
+``source_namespace``      The namespace of the pod if the flow source is from a Kubernetes pod.
+``source_pod``            The pod name of the pod if the flow source is from a Kubernetes pod.
+``source_workload``       The name of the source pod's workload (Deployment, Statefulset, Daemonset, ReplicationController).
```
Yeah, and a CronJob too actually. I'll update it. I'm not sure if I should mention DeploymentConfigs for openshift or not, but I could list that too.

chancez replied:
Updated to mention these.

michi-covalent reviewed (approved):
🚢

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -496,9 +496,18 @@ and the ``http`` metric with the ``destinationContext=pod-short`` option.
+
+- ``sourceContext`` - Configures the ``source`` label on metrics.
+- ``destinationContext`` - Configures the ``destination`` label on metrics.
+- ``labelsContext`` - Configures a list labels to be enabled on metrics.
```
```suggestion
- ``labelsContext`` - Configures a list of labels to be enabled on metrics.
```

chancez replied:
Fixed

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -530,6 +539,29 @@ and use the IP address of the target instead.
+========================= ===============================================================================
+``source_namespace``      The namespace of the pod if the flow source is from a Kubernetes pod.
+``source_pod``            The pod name of the pod if the flow source is from a Kubernetes pod.
+``source_workload``       The name of the source pod's workload (Deployment, Statefulset, Daemonset, ReplicationController, CronJob, Job, DeploymentConfig (Openshift)).
```
```suggestion
``source_workload``       The name of the source pod's workload (Deployment, Statefulset, Daemonset, ReplicationController, CronJob, Job, DeploymentConfig (OpenShift)).
```

Here and below. And then I _think_ you can remove Openshift from the spelling_wordlist.txt, because it appears elsewhere spelled as “OpenShift” in the docs without triggering spelling errors.

chancez replied:
Fixed

qmonnet commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -129,30 +159,123 @@ func parseContext(s string) (cs ContextIdentifierList, err error) {
+					labelValue = "unknown"
+				}
+			default:
+				// Label is in contextLabelsList but isn't handled the switch
```
```suggestion
				// Label is in contextLabelsList but isn't handled in the switch
```

chancez replied:
Fixed

qmonnet commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -281,10 +404,26 @@ func destinationIPContext(flow *pb.Flow) (context string) {
 
+func getK8sAppFromLabels(labels []string) string {
+	ls := ciliumLabels.ParseLabelArrayFromArray(labels)
+	// iterate over the set of pod app labels we source from and return the first
```
```suggestion
	// Iterate over the set of pod app labels we source from and return the first
```

chancez replied:
I'll fix this, but technically I only moved this function ;)

chancez replied:
Fixed

qmonnet commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -345,6 +484,16 @@ func (o *ContextOptions) GetLabelValues(flow *pb.Flow) (labels []string) {
 func (o *ContextOptions) GetLabelNames() (labels []string) {
+	if len(o.Labels) != 0 {
+		// We must iterate over contextLabelsList to ensure the order of the label
+		// names the same order as label values in GetLabelValues
```
```suggestion
		// names the same order as label values in GetLabelValues.
```

chancez replied:
Fixed

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -530,6 +539,29 @@ and use the IP address of the target instead.
+When specifying the flow context, multiple values can be specified by separating them via the ``,`` symbol.
+All labels listed are included in the metric, even if empty. For example, a metric configuration of
+``http:labelsContext=source_namespace,source_pod`` will add the ``source_namespace`` and ``source_pod``
+labels to all Hubble http metrics.
```
```suggestion
labels to all Hubble HTTP metrics.
```

chancez replied:
Fixed

qmonnet reviewed (changes requested):
Apologies for the review delay.

Looks good, with a few minor nitpicks (inline below).

Also a few typos in the commit description: “This PR” -> “In this PR” (or even “In this commit”, Git logs have no notion of PRs), “illistrates” -> “illustrates”.

kaworu commented on Documentation/observability/metrics.rst:
```diff
@@ -530,6 +539,29 @@ and use the IP address of the target instead.
+Option Value              Description
+========================= ===============================================================================
+``source_namespace``      The namespace of the pod if the flow source is from a Kubernetes pod.
+``source_pod``            The pod name of the pod if the flow source is from a Kubernetes pod.
```
"pod name of the pod" sounds a bit redundant here?

chancez replied:
Updated

kaworu commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -129,30 +159,123 @@ func parseContext(s string) (cs ContextIdentifierList, err error) {
+			case "source_pod":
+				if flow.GetSource() != nil {
+					labelValue = flow.GetSource().PodName
+				}
```
nit: protobuf methods are `nil` safe so you can simplify here
```suggestion
				labelValue = flow.GetSource().GetPodName()
```
ditto for the other branches of the `switch`.

chancez replied:
Ah, I copied this approach from the code we already had in here, so I thought it was necessary. Ill try that.

chancez replied:
Updated

kaworu commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -129,30 +159,123 @@ func parseContext(s string) (cs ContextIdentifierList, err error) {
+			default:
+				// Label is in contextLabelsList but isn't handled the switch
+				// statement. Programmer error.
+				panic(fmt.Sprintf("label %s not mapped in labelsContext, please update labelsContext", label))
```
Even though we have a bug if we reach this codepath, it's does not seems critical enough to `panic` the Cilium agent.

chancez replied:
Hmm I see your point but, these functions don't return errors so I wasn't sure of a better better approach without refactoring GetLabelValues to return an error, which means all plugin handlers will have to handle that. The other option is that we turn these into some kind of method and provide a logger to the object so it can log the error.

In any case, if we don't produce the right number of output labels, the prometheus instrumentation will throw an error since they labels values list won't match the length of the labels list.

I guess if prometheus will error, there's no point in the logger approach, so I think returning the error would be best in this case, even though it should never be hit unless we have a bug.

Another thought is we could add a `recover()` around the hubble server start and catch hubble panics if we don't want hubble bugs to crash cilium (I guess this would be similar to just logging it though).

chancez replied:
Actually, even if we return the error, metrics plugin `ProcessFlow` handlers don't return errors either, and none of the handlers get plumbed a logger either.

chancez replied:
Ok, added a new commit to replace the panic. It involved some refactoring so I opted to separate that commit out from the rest.

kaworu reviewed (dismissed):
Thanks for the PR @chancez! Couple of nits, but more importantly `panic`'ing doesn't seem appropriate to me in this patch.

chancez committed:
hubble/metrics: Replace panic in contextLabels with error log

Instead of panicing when contextLabels doesn't handle a defined label,
return the error up the stack and log it.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

kaworu commented on pkg/hubble/metrics/api/api.go:
```diff
@@ -63,18 +64,25 @@ type Handler interface {
+			// Continue running the remaining metrics handlers, since one failing
+			// shouldn't impact the other metrics handlers.
+			processingErr = multierr.Append(processingErr, err)
+		}
```
nit: we don't necessarily need to check explicitly for err, as `multierr.Append` support either (or both) arguments to be `nil`

```suggestion
		err := mh.ProcessFlow(ctx, flow)
		processingErr = multierr.Append(processingErr, err)
		// Continue running the remaining metrics handlers, since one failing
		// shouldn't impact the other metrics handlers.
```

chancez replied:
Fixed

kaworu commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -129,30 +159,123 @@ func parseContext(s string) (cs ContextIdentifierList, err error) {
+			default:
+				// Label is in contextLabelsList but isn't handled in the switch
+				// statement. Programmer error.
+				return nil, fmt.Errorf("label %s not mapped in labelsContext, please update labelsContext", label)
```
If I understand correctly, we are asking a develop to update the `labelsContext` `func` implementation (specifically this `switch` statement) and not the user-facing `labelsContext` metric context option. Is that unambiguous? If not, maybe we could clarify that this is a bug here rather than a misuse. Does that make sense?

gandro replied:
I agree with Alex. The way we do this in other places is to explicitly ask users to report this to Cilium as a bug. For example:
https://github.com/cilium/cilium/blob/af21e0380a19826823f803f400d24aefae28d56c/daemon/cmd/datapath.go#L304

Let's use a similar wording here.

chancez replied:
Fixed

kaworu commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -360,6 +513,12 @@ func (o *ContextOptions) GetLabelNames() (labels []string) {
+	if len(o.Labels) != 0 {
+		status = append(status, "labels="+o.Labels.String())
+
+		// status = append(status, labelsContext(o.Labels, flow)...)
```
do we need to keep this?

chancez replied:
Fixed

kaworu reviewed (commented):
Thanks for the update @chancez, couple more nits / comments.

gandro commented on Documentation/observability/metrics.rst:
```diff
@@ -530,6 +539,29 @@ and use the IP address of the target instead.
+``destination_pod``       The pod name if the flow destination is from a Kubernetes pod.
+``destination_workload``  The name of the destination pod's workload (Deployment, Statefulset, Daemonset, ReplicationController, CronJob, Job, DeploymentConfig (OpenShift)).
+``destination_app``       The app name of the source pod, derived from pod labels (``app.kubernetes.io/name``, ``k8s-app``, or ``app``).
+``reporter``              Identifies the origin of the flow/request. It is set to ``source`` if it originated from the client, and ``destination`` if it originated from the server.
```
This documentation for `reporter` does not seem to match the implementation. But I do like `source` and `destination` better than `client` and `server`.

chancez replied:
Ah, I originally used source/destination, but I switched to `client/server` which is inline with the OTEL semantics conventions (uses client/server terminology).

chancez replied:
I updated the documentation to reflect the usage of `client/server`. If you think we should use source/destination, we could also consider adding another label option such as `traffic_direction` which just exposes the value as `egress|ingress|unknown`, even though it overlaps with `reporter`. Might be more useful for non-HTTP flows where the client/server terminology might be less correct.

Another option is to remove `reporter` from the contextOptions and to put it directly on the HTTP metrics labels themselves. I'm actually thinking this might be best.

chancez replied:
Ok so I ended up doing this:
- Replace `reporter` contextLabel option with `traffic_direction`, which has label values `ingress|egress|unknown`. 
- Add `reporter` label directly to HTTP metrics with values `client|server|unknown`. 

This is also great because it ensures the `reporter` label is present on HTTP metrics when no contextLabel options are configured, which means it's more likely that users pay attention to it when writing queries, which should help with preventing people from double counting when using egress/ingress policies on the same port.

chancez replied:
Only if you enable it. It's two ways of saying the same thing, but in the context of `client|server` I think that doesn't always make sense for all flows, so you may use `reporter` for http, and `traffic_direction` for say: flow metrics.

chancez replied:
Information is only duplicated if you explicitly enable `traffic_direction`, I think it's not a huge deal here.

gandro commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -281,10 +404,30 @@ func destinationIPContext(flow *pb.Flow) (context string) {
 }
 
+func getK8sAppFromLabels(labels []string) string {
+	ls := ciliumLabels.ParseLabelArrayFromArray(labels)
```
This is unfortunately a bit of an expensive operation, and we've had to optimize similar code in the past.

Have you considered doing a `strings.HasPrefix` on `podAppLabels` and only parse the label if the prefix matches?

chancez replied:
I figured re-using the code we already had was best. Is there not any less-expensive option available in the labels package we can use?

gandro replied:
I don't think there is an existing function for this. Hubble's use of `[]string` predates the integration with the Cilium codebase, which typically uses `labels.LabelArray` as the in-memory type.

chancez replied:
Ok I've rewritten that function in a new commit. I dislike nested loops but I believe this is still probably more efficient, as it always had to loop over all the flow labels, which it still does, and the inner loop is over the app labels, and it's size is constant.

gandro reviewed (commented):
Looks good to me overall. A few minor items, besides to what Alex already pointed out.

chancez committed:
hubble/metrics: Use less expensive check for app labels

ParseLabelArrayFromArray can be expensive, so instead, use a more
optimized code path that uses strings.HasPrefix to check if the label
key is correct, then only calls ParseLabel after.

Nested loops are often bad, but in this case, the inner loop is a fixed
size, and should be quite fast. We can't use a map because we want to
use specific app labels before other app labels if they are present and
non-empty.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
hubble/metrics: Add reporter label to HTTP metrics

Also adds documentation on each HTTP label, and explains the new
reporter label.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

gandro commented on pkg/hubble/metrics/api/context.go:
```diff
@@ -281,10 +400,32 @@ func destinationIPContext(flow *pb.Flow) (context string) {
+func getK8sAppFromLabels(labels []string) string {
+	for _, label := range labels {
+		for _, appLabel := range podAppLabels {
+			if strings.HasPrefix(label, appLabel) {
```
I think you want to check for `strings.HasPrefix(label, appLabel+"=")`

chancez replied:
Ah good catch. Fixed.
